compiler/rustc_typeck/src/collect.rs

```diff
@@ -984,54 +1030,132 @@ fn adt_def(tcx: TyCtxt<'_>, def_id: DefId) -> &ty::AdtDef {
 /// the transitive super-predicates are converted.
 fn super_predicates_of(tcx: TyCtxt<'_>, trait_def_id: DefId) -> ty::GenericPredicates<'_> {
     debug!("super_predicates(trait_def_id={:?})", trait_def_id);
-    let trait_hir_id = tcx.hir().local_def_id_to_hir_id(trait_def_id.expect_local());
+    tcx.super_predicates_that_define_assoc_type((trait_def_id, None))
+}
 
-    let item = match tcx.hir().get(trait_hir_id) {
-        Node::Item(item) => item,
-        _ => bug!("trait_node_id {} is not an item", trait_hir_id),
-    };
+/// Ensures that the super-predicates of the trait with a `DefId`
+/// of `trait_def_id` are converted and stored. This also ensures that
+/// the transitive super-predicates are converted.
+fn super_predicates_that_define_assoc_type(
```

Comment:
Hmm so this is a query... it's a bit unusual in that most query providers *either* operate on local def-ids *or* external def-ids but not both. (The typical strategy, in other words, would be to define this query by encoding it into the metadata.) But that seems kind of silly here, I don't see a reason to encode more into the metadata. So this setup is probably ok. But I feel like we have to be sure to setup the provider also for external crates or we will see ICEs -- I'm a bit surprised our tests aren't encountering them yet, actually.